scripts/testramfs/testramfs.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/u-root/u-root/pkg/cpio"
+	_ "github.com/u-root/u-root/pkg/cpio/newc"
+)
+
+func main() {
+	flag.Parse()
+
+	if flag.NArg() != 1 {
+		log.Fatalln("usage: %s <cpio-path>", os.Args[0])
+	}
+
+	if err := Test(flag.Args()[0]); err != nil {
+		log.Fatalln(err)
+	}
+}
+
+func Test(name string) error {
+	// So, what's the plan here?
+	//
+	// - new mount namespace
+	//   - root mount is a tmpfs mount filled with the archive.
+	//
+	// - new PID namespace
+	//   - archive/init actually runs as PID 1.
+
+	// Whatever setup we do shouldn't affect the caller.
+	if err := syscall.Unshare(syscall.CLONE_NEWNS); err != nil {
+		return fmt.Errorf("unshare: %v", err)
+	}
+
+	tempDir, err := ioutil.TempDir("", "u-root")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+	if err := syscall.Mount("", tempDir, "tmpfs", 0, ""); err != nil {
```

Comment:
The default behaviour of private vs. shared depends on whether you are running systemd. It would be good to set it back to private.

See: https://github.com/golang/go/issues/19661

Reply:
The only safe way to do this right is to restart yourself with cmd.Unshare set. That will do the right thing.